General optimizations

pnwparksfan opened the pull request.

TheMaybeast committed:
Optimizes ControlsInput registering

TheMaybeast committed:
Simplify Modes and Control Groups logic

TheMaybeast committed:
Remove Core.Lights namespace

TheMaybeast committed:
General code cleanup

pnwparksfan commented on DLSv2/Core/Wrappers.cs:
```diff
@@ -0,0 +1,162 @@
+                        {
+                            Enabled = true;
+                            Index = lightCG.BaseControlGroup.Modes.Count - 1;
+                            return;
+                        }
+                    }
+                    if (newIndex < 0)
+                    {
+                        Enabled = false;
+                        Index = 0;
+                    }
+                    else
+                    {
+                        Enabled = true;
+                        Index = newIndex;
+                    }
+                    break;
+                }
+                case ControlGroupInstance<AudioControlGroup> audioCG:
+                    if (Enabled == false)
+                    {
+                        if (Index == 0)
+                        {
+                            Enabled = true;
+                            Index = audioCG.BaseControlGroup.Modes.Count - 1;
+                            return;
+                        }
+                    }
+                    if (newIndex < 0)
+                    {
+                        Enabled = !cycleOnly;
+                        Index = cycleOnly ? 0 : audioCG.BaseControlGroup.Modes.Count - 1;
+                    }
+                    else
+                    {
+                        Enabled = true;
+                        Index = newIndex;
+                    }
+                    break;
+            }
```
This logic seems to be the same for audio control groups and light control groups. Likewise in MoveToNext. It might make sense to have a base ControlGroup class that both light and audio control groups inherit from. Then ControlGroupInstance can have `<T>` be forced to use a `BaseControlGroup` (`ControlGroupInstance<T> where T : BaseControlGroup` so you can use properties of `T` in methods like next/prev.

TheMaybeast replied:
Audio and Light control groups differ on what happens on if newIndex < 0.

pnwparksfan replied:
You could still have the same base class so the types are cleaner, and then implement a different logic for just that line

pnwparksfan commented on DLSv2/Core/Wrappers.cs:
```diff
@@ -0,0 +1,162 @@
+﻿namespace DLSv2.Core
+{
+    public class ModeInstance<T>
+    {
+        public T BaseMode { get; }
+        // General
+        public bool Enabled;
+        public bool EnabledByTrigger;
+
+        public ModeInstance(T mode)
+        {
+            BaseMode = mode;
+            Enabled = false;
+            EnabledByTrigger = false;
+        }
+    }
```
Consider if making light/audio Modes share a base class also makes sense.

pnwparksfan commented on DLSv2/Core/ManagedVehicle.cs:
```diff
@@ -324,22 +324,22 @@ public void RegisterInputs()
                     // Clears audio control groups
-                    foreach (string key in AudioControlGroups.Keys.ToList())
-                        AudioControlGroups[key] = (false, 0);
+                    foreach (var cG in AudioControlGroups.Values.ToList())
```
Do these all need to have `.ToList()`? I think the `foreach` should work fine directly on a `Values` collection.

TheMaybeast replied:
IIRC this is in so I can set the value within the collection while looping.

pnwparksfan replied:
If not changing the keys and only changing the values, it should be safe to loop over the keys directly and set the corresponding values.

pnwparksfan commented on DLSv2/Core/ManagedVehicle.cs:
```diff
@@ -127,16 +125,15 @@ public bool LightsOn
+            ControlsManager.ClearInputs();
             // Light Inputs
-            foreach (ControlGroup cG in ControlGroupManager.ControlGroups[Vehicle.Model].Values)
+            foreach (ControlGroup cG in LightControlGroups.Values.Select(x => x.BaseControlGroup))
```
I would change this to do `foreach(ControlGroupWrapper cgw in LightControlGroups.Values)`. You can get the base control group where you need it, but this would allow you to avoid all of the repeated use of `LightControlGroups[cg.Name].Toggle()` etc. and just directly use `cgw.Toggle()`. You're basically getting the items from a dictionary, looping through just the name of them, then getting them back from the same dictionary, which is inefficient when you could just use the items directly.

pnwparksfan commented on DLSv2/Core/ManagedVehicle.cs:
```diff
@@ -192,17 +189,17 @@ public void RegisterInputs()
 
             // Audio Control group and modes keys
-            foreach (AudioControlGroup cG in AudioControlGroupManager.ControlGroups[Vehicle.Model].Values)
+            foreach (AudioControlGroup cG in AudioControlGroups.Values.Select(x => x.BaseControlGroup))
```
Same thing here: loop through the wrappers, not the base control groups, then use the wrappers directly

pnwparksfan commented on DLSv2/Utils/Loaders.cs:
```diff
@@ -16,39 +13,32 @@ internal class Loaders
                     }
 
-                    string name = Path.GetFileNameWithoutExtension(file);
+                    var name = Path.GetFileNameWithoutExtension(file);
```
Why did you change so many things to use `var`? It's valid, but sometimes it's confusing. 

When you have something like `StreamReader reader = new StreamReader(file)`, it makes sense to use `var` because it's very obviously a `StreamReader` because the constructor is right there. But when you have an arbitrary function whose return type is not necessarily obvious like `Path.GetFileNameWithoutExtension(file)`, you don't necessarily know it'll return a `string`, so I think it's more clear to explicitly define the type as `string`.

pnwparksfan commented on DLSv2/Utils/VehicleExtensions.cs:
```diff
@@ -129,5 +129,36 @@ public static bool IsBumperBroken(this Vehicle vehicle, bool front, bool detache
             return NativeFunction.CallByName<bool>(nativeName, vehicle, front);
         }
+
+        public static void SetIndicator(this Vehicle vehicle, VehicleIndicatorLightsStatus indStatus)
+        {
+            if (!vehicle) return;
+
+            switch (indStatus)
+            {
+                case VehicleIndicatorLightsStatus.Off:
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 0, false);
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 1, false);
+                    break;
+                case VehicleIndicatorLightsStatus.LeftOnly:
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 0, false);
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 1, true);
+                    break;
+                case VehicleIndicatorLightsStatus.RightOnly:
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 0, true);
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 1, false);
+                    break;
+                case VehicleIndicatorLightsStatus.Both:
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 0, true);
+                    NativeFunction.Natives.SET_VEHICLE_INDICATOR_LIGHTS(vehicle, 1, true);
+                    break;
+            }
+        }
```
Why use this extension method instead of `Rage.Vehicle.IndicatorLightsStatus`?

pnwparksfan commented on DLSv2/Utils/VehicleExtensions.cs:
```diff
@@ -129,5 +129,36 @@ public static bool IsBumperBroken(this Vehicle vehicle, bool front, bool detache
+            }
+        }
+
+        public static void SetInteriorLight(this Vehicle vehicle, bool isOn)
+        {
+            if (!vehicle) return;
+            vehicle.IsInteriorLightOn = isOn;
+        }
```
Also why this rather than setting `IsInteriorLightOn` directly?

pnwparksfan commented on DLSv2/Utils/DLSExtensions.cs:
```diff
@@ -121,7 +121,121 @@ internal static SirenSetting GetDefaultSirenSetting(this Vehicle veh)
+            managedVehicle.Vehicle.EmergencyLightingOverride = eL;
+        }
+
+        /*internal static void DebugCurrentModes(this Vehicle vehicle)
```
Is this disabled just because things changed so the logging doesn't work currently? Seems like it will be very helpful for testing if the changes work right. Do you plan to add it back?

TheMaybeast replied:
Yeah, disabled due to the restructuring. Will work it back on before merging.

pnwparksfan commented on DLSv2/Core/ManagedVehicle.cs:
```diff
@@ -127,16 +125,15 @@ public bool LightsOn
         // Registers input
         public void RegisterInputs()
         {
+            ControlsManager.ClearInputs();
```
I still don't love re-registering all inputs every time the managed vehicle changes. See comment from DMs: 

> Registering/unregistering inputs every time you switch active vehicles is inefficient and confusing. I recommend registering all inputs from the INI once when the plugin loads. When the current "active" player-owned vehicle changes, unsubscribe all of the previous vehicle's event handlers and subscribe to all of the new vehicle's input event handlers, but don't re-load all inputs from scratch. Alternatively, never unsubscribe the event handlers, and just check the new bool for is it the player's active vehicle and return out of the handler if it isn't.

pnwparksfan reviewed (commented):
Left some comments on the changes below, feel free to discuss further via DMs

TheMaybeast committed:
Fix broken siren audio

TheMaybeast committed:
Removes unnecessary ToList calls

TheMaybeast committed:
Optimizes input registration

TheMaybeast committed:
Uses RPH indicator method instead of natives

TheMaybeast committed:
Move away from Indicator extension

TheMaybeast committed:
Move away from Interior Light extension

TheMaybeast committed:
Merge branch 'main' into general-optimizations

TheMaybeast committed:
Fixes LightMode condition

TheMaybeast committed:
Clears key on leave vehicle

TheMaybeast committed:
Manually executes ExtraRepairPatch

TheMaybeast committed:
Better secure ExtraRepairPatch

TheMaybeast committed:
Fixes Mode Toggle

TheMaybeast committed:
Removes extra patch on unload.

TheMaybeast committed:
Fixes AI lights not enabling

TheMaybeast committed:
Merges Light and Audio mode/CG logic

TheMaybeast committed:
Merge branch 'main' into general-optimizations

TheMaybeast committed:
Merge branch 'main' into go-inheritance

TheMaybeast committed:
Relocate build binaries

pnwparksfan committed:
Created wrapper classes to avoid excessive generics

TheMaybeast committed:
Merge pull request #19 from TheMaybeast/go-inheritance

[General Optimizations] Inheritance for Modes and Control Groups

TheMaybeast committed:
Multiple changes

- Updates to latest C# version.
- Fixes parsing XML (inheritance broke it);
- Moves away from clunky clear sirens to proper memory patching;
- Moves memory functions/offsets/structs to their own namespace;
- Moves all memory reads/writes to Marshal for safety, except on Memory.Unsafe.

TheMaybeast committed:
Removes unnecessary logging spam

TheMaybeast committed:
Updates logger

TheMaybeast committed:
Moves version check and "LOADED DLS" message.

Also moves unnecessary logging to DEBUG only.

TheMaybeast committed:
Fixes Debug Commands

TheMaybeast committed:
Makes DebugCurrentMode output to in-game console

TheMaybeast committed:
Fixes log not saving after exit

pnwparksfan committed:
Fixed console output for nested condition status in debug command

pnwparksfan committed:
Multiple fixes/updates:

 - Added debug command option to show/hide verbose conditions details
 - Change default value for exclusive on control groups to true
 - Fixed Yield implementation to allow overwriting when specified,
   for either enabled or disabled values, and corrected code and example
   XML accordingly
 - Optimized CachedGameTime by implementation of a separate fiber

pnwparksfan committed:
Adjust requirement event handler to re-enable modes

pnwparksfan committed:
Misc code cleanup and optimization

pnwparksfan committed:
Separated vehicle player owner logic into separate class, added event handler for changed ownership

pnwparksfan committed:
Fixed default mode behavior and dynamic switching between NPC/player owned vehicle
